### PR TITLE
Add Leader election to backup-restore and allow only backup-restore Leader to take and uploads the snapshots.

### LINKED ISSUE #321: [Feature] Add leader-election to the backup-restore sidecar according to the multi-node ETCD proposal.
**Feature (What you would like to be added):**

Add leader-election to the backup-restore sidecar according to the multi-node ETCD proposal. Bring the [snapshotters](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#backup) (both full and incremental) and the [defragmentation](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#defragmentation) control loops/go routines under the control of the leader-election as described [here](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#work-flows-in-etcd-backup-restore).

Also, introduce the logic to regularly cleanup superfluous ETCD members which do not have a corresponding `Pod` in the `StatefulSet` as mentioned in https://github.com/gardener/etcd-druid/pull/189.

**Motivation (Why is this needed?):**

Pick individually executable pieces of the [multi-node proposal](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md).

**Approach/Hint to the implement solution (optional):**

The preferred leader-election mechanism (because of the implementation and operational simplicity) is based on [the leading etcd member](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#leading-etcd-main-containers-sidecar-is-the-backup-leader). This can be done using the same heart-beat on the corresponding ETCD member's [endpoint status](https://github.com/etcd-io/etcd/blob/f82b5cb7768dacad9fb310232c1383b4e6718378/client/v3/maintenance.go#L53) as mentioned in #333

### LINKED ISSUE #354: [Feature]To take out of sync full/delta snapshot,  delegate the incoming http request from backup Follower to the backup Leader.
**Feature (What you would like to be added):**
If backup-restore `Follower` receives the HTTP/HTTPS  requests to take [out-of-sync full/delta snapshot](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/server/httpAPI.go#L103-L105) then delegate that incoming request to the backup-restore `Leader` and Let the backup-restore leader take/upload the out-of-sync snapshot(full as well as incremental).

**Motivation (Why is this needed?):**
While adding support to bring the [snapshotters](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#backup) (both full and incremental) go routines under the control of the leader-election as described [here](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#work-flows-in-etcd-backup-restore), We observed that all backup-restore Follower's [http handler](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/server/backuprestoreserver.go#L93-L105) don't have updated [SnapshotterState](https://github.com/gardener/etcd-backup-restore/blob/13ddcaa7ec25be4b23f204124ac3f718fe9920c9/pkg/snapshot/snapshotter/snapshotter.go#L68-L92), if the request comes to the REST endpoints fo `follower` to take out of sync full/delta snapshot then the snapshot taken will not be the latest, and it is not good to let backup-restore `Follower` take the snapshot, it should delegate the request to backup-restore `Leader`.

**Approach/Hint to the implement solution (optional):**

### PR DESCRIPTION
**Pull request by ishan16696**
**What this PR does / why we need it**:
Added leader-election to the backup-restore sidecar.

Added support to bring the snapshotters (both full and incremental)  and defragmentation go routines under the control of the leader-election.

To take out of sync full/delta snapshot and to get the latest snapshot metadata, delegate the incoming HTTP/HTTPS request from `backup Follower` to the `backup Leader`.

**Which issue(s) this PR fixes**:
Fixes #321 Part-1
Fixes #354 

**Special notes for your reviewer**:

**Release note**:
```improvement operator
Added the leader-election to the backup-restore. Added support to allow only backup-restore Leader to take/upload the snapshots and trigger the defragmentation.
```

### CODE REVIEW

File: pkg/server/backuprestoreserver.go [hunk 1]
```diff
@@ -151,19 +153,47 @@ func (b *BackupRestoreServer) runServerWithSnapshotter(ctx context.Context, rest
 	handler := b.startHTTPServer(etcdInitializer, ssr)
 	defer handler.Stop()
 
-	ssrStopCh := make(chan struct{})
+	leaderCallbacks := &leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(ctx context.Context) {
+			// Get the updated snapshotter state
+			err := ssr.GetLatestSnapshotterState()
```
Comment (amshuman-kr) [inline pkg/server/backuprestoreserver.go:159]: This function call will race with the snapshotter goroutine.

File: pkg/server/backuprestoreserver.go [hunk 2]
```diff
@@ -151,19 +153,47 @@ func (b *BackupRestoreServer) runServerWithSnapshotter(ctx context.Context, rest
...
+			if err != nil {
+				fmt.Printf("Unable to get Latest SnapshotterState: %v", err)
```
Comment (amshuman-kr) [inline pkg/server/backuprestoreserver.go:161]: In this case, we have a leader election and snapshotter is not updated with latest state. The next time it will be rectified is when the leader election changes next time. What happens in the meantime?

File: pkg/server/backuprestoreserver.go [hunk 3]
```diff
@@ -151,19 +153,47 @@ func (b *BackupRestoreServer) runServerWithSnapshotter(ctx context.Context, rest
...
+			}
+			handler.Snapshotter = ssr
```
Comment (amshuman-kr) [inline pkg/server/backuprestoreserver.go:163]: This will race with the HTTP server goroutine.

File: pkg/leaderelection/leaderelection.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+		StopCh:               make(chan struct{}),
+		ProbeEtcdch:          make(chan struct{}),
+		PrevElectionResult:   ElectionLost,
+		Config:               leaderElectionConfig,
+		Callbacks:            callbacks,
+	}, nil
+}
+
+// StartElection starts the LeaderElection to Elect the backup-restore's Leader.
+func (le *LeaderElector) StartElection(ctx context.Context) {
```
Comment (amshuman-kr) [inline pkg/leaderelection/leaderelection.go:44]:
```suggestion
// Run starts the LeaderElection to Elect the backup-restore's Leader.
func (le *LeaderElector) Run(ctx context.Context) {
```

`StartElection` means one election. But here it is a continuous series of elections. So `Run` is a better name for the function.

File: pkg/server/backuprestoreserver.go [hunk 5]
```diff
@@ -177,6 +207,9 @@ func (b *BackupRestoreServer) runEtcdProbeLoopWithSnapshotter(ctx context.Contex
 			return
 		default:
 			err = b.probeEtcd(ctx)
+			if err == nil {
+				<-le.ProbeEtcdch
+			}
```
Comment (amshuman-kr) [inline pkg/server/backuprestoreserver.go:212]: Why is this needed when we already have callbacks?

File: pkg/server/backuprestoreserver.go [hunk 6]
```diff
@@ -151,19 +153,47 @@ func (b *BackupRestoreServer) runServerWithSnapshotter(ctx context.Context, rest
...
+		},
+		OnStoppedLeading: func() {
+			// stops the running snapshotter
+			ssr.SsrStateMutex.Lock()
+			if ssr.SsrState == brtypes.SnapshotterActive {
+				ssr.SsrStateMutex.Unlock()
+				ssrStopCh <- emptyStruct
+			} else {
+				ssr.SsrStateMutex.Unlock()
+			}
```
Comment (amshuman-kr) [inline pkg/server/backuprestoreserver.go:173]:
```suggestion
			ssr.SsrStateMutex.Lock()
			ssr.SsrStateMutex.Unlock()
			if ssr.SsrState == brtypes.SnapshotterActive {
				ssrStopCh <- emptyStruct
			}
```

This might inactivate the snapshotter but what happens if some other goroutine activates it (by toggling using the same `sstStopCh`) before the leader election has been won?

File: pkg/server/types.go [hunk 7]
```diff
@@ -34,6 +35,7 @@ type BackupRestoreComponentConfig struct {
 	CompressionConfig       *compressor.CompressionConfig  `json:"compressionConfig,omitempty"`
 	RestorationConfig       *brtypes.RestorationConfig     `json:"restorationConfig,omitempty"`
 	DefragmentationSchedule string                         `json:"defragmentationSchedule"`
+	LeaderElectionConfig    *leaderelection.ElectionConfig `json:"electionConfig,omitempty"`
```
Comment (amshuman-kr) [inline pkg/server/types.go:38]:
```suggestion
	LeaderElectionConfig    *leaderelection.Config `json:"leaderElectionConfig,omitempty"`
```

File: pkg/snapshot/snapshotter/snapshotter.go [hunk 8]
```diff
@@ -99,10 +99,35 @@ func NewSnapshotter(logger *logrus.Entry, config *brtypes.SnapshotterConfig, sto
...
+
+	err = ssr.GetLatestSnapshotterState()
 	if err != nil {
 		return nil, err
+	}
+	return ssr, nil
+}
+
+// GetLatestSnapshotterState load the previous snapshot state for ssr.
+func (ssr *Snapshotter) GetLatestSnapshotterState() error {
```
Comment (amshuman-kr) [inline pkg/snapshot/snapshotter/snapshotter.go:126]: This function gets called from at least two goroutines and therefore will race.

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Thanks for the first-cut implementation @ishan16696. I have some comments and questions below.

File: pkg/leaderelection/leaderelection.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+// NewLeaderElector returns LeaderElector configurations.
+func NewLeaderElector(logger *logrus.Entry, etcdConnectionConfig *etcdutil.EtcdConnectionConfig, leaderElectionConfig *ElectionConfig, callbacks *LeaderCallbacks) (*LeaderElector, error) {
+	return &LeaderElector{
+		logger:               logger.WithField("actor", "leader-elector"),
+		EtcdConnectionConfig: etcdConnectionConfig,
+		CurrentState:         DefaultCurrentState,
+		StopCh:               make(chan struct{}),
+		ProbeEtcdch:          make(chan struct{}),
+		PrevElectionResult:   ElectionLost,
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:37]: We need to take certain actions when we observe a new leader(if `PrevElectionResult =ElectionLost`) and we have to do nothing when a particular backup-restore keeps winning the election(if `PrevElectionResult =ElectionWon`)

**Comment by ishan16696** [comment]
I have also tested the cluster-based Integration test locally. Logs: Please find attached.
[clusterIntegrationTest.log](https://github.com/gardener/etcd-backup-restore/files/6982620/clusterIntegrationTest.log)

**Review by stoyanr — CHANGES_REQUESTED** [review]
As discussed, the PR seems to not introduce any regressions for the single member case. For the multi-member case, the owner check watchdog would ensure the proper terminating of the connections to the old leader, and keeping the new leader unavailable, which should work well with the current implementation since it does not allow connections to follower replicas. It should be double-checked however if this is really what we want.

In addition, I posted some feedback / comments on other aspects of the PR.

File: pkg/leaderelection/leaderelection.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+	client, err := etcdutil.GetTLSClientForEtcd(le.EtcdConnectionConfig)
+	if err != nil {
+		return false, &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to create etcd client: %v", err),
+		}
+	}
+	defer client.Close()
+
+	if len(le.EtcdConnectionConfig.Endpoints) > 0 {
+		endPoint = le.EtcdConnectionConfig.Endpoints[0]
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:122]: It is the endpoint of `etcd` which comes fom the [etcdConnectionConfig](https://github.com/gardener/etcd-backup-restore/blob/5dfcc1f848a9f325d41a24eae4defb70d997c215/pkg/etcdutil/types.go#L35-L49), I think it should be passed through environment variable but I'm also not so sure, so this thing need to be take care.

cc @abdasgupta @shreyas-s-rao @timuthy @aaronfern

File: pkg/leaderelection/leaderelection.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+		case <-time.After(le.Config.ReelectionPeriod.Duration):
+			isLeader, err := le.IsLeader(ctx)
+			if err != nil {
+				le.logger.Errorf("Failed to elect the backup-restore leader: %v", err)
+
+				// set the CurrentState of backup-restore.
+				// stops the Running Snapshotter.
+				// wait for Reelection to happen.
+				le.CurrentState = StateUnknown
+				le.logger.Infof("backup-restore is in: %v", le.CurrentState)
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:63]: > backup-restore is redundant

agreed , If I changed the logs to `Current leader election state is`
logs would be looks like this:
```
Current leader election state is Leader  actor=leader-elector
Current leader election state changed to Follower actor=leader-elector
Current leader election state is Follower.  actor=leader-elector
Current leader election state is UnkownState   actor=leader-elector
```
I don't like this either, I will try to come up with nice msg.

File: pkg/leaderelection/leaderelection.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// IsLeader checks whether the current backup-restore is leader or not.
+func (le *LeaderElector) IsLeader(ctx context.Context) (bool, error) {
+	le.logger.Info("checking the leadershipStatus...")
+	var endPoint string
+	client, err := etcdutil.GetTLSClientForEtcd(le.EtcdConnectionConfig)
+	if err != nil {
+		return false, &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to create etcd client: %v", err),
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:116]: thanks for suggestion, I will update

**Review by abdasgupta — COMMENTED** [review]
Thanks for the nice PR. Added few comments

File: example/00-backup-restore-server-config.yaml [hunk 13]
```diff
@@ -51,3 +51,14 @@ defragmentationSchedule: "0 0 */3 * *"
...
+
+leaderElectionConfig:
+  reelectionPeriod: "5s"
+  etcdConnectionTimeout: "5s"
+
+healthConfig:
+  snapshotLeaseRenewalEnabled: false
+  memberLeaseRenewalEnabled: false
+  heartbeatDuration: "30s"
+  fullSnapshotLeaseName: "full-snapshot-revisions"
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply example/00-backup-restore-server-config.yaml:63]: I take the what json is defined: Can you check this [here](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/types/health.go#L41-L47)

File: pkg/leaderelection/leaderelection.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:135]: yeah, endpoints of etcd comes as an array but actually Question was "How to make sure that endpoints present at 0th index in a array is always endpoint of their corresponding etcd?" @amshuman-kr also raised this question.
answer to this questions could be that etcd endpoints should passed as environment variable to their corresponding sidecar.

File: pkg/compactor/compactor_suite_test.go [hunk 14]
```diff
@@ -65,7 +69,7 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	logger.Infof("Snapshot Directory is: %s", testSnapshotDir)
 
 	// Start the main ETCD process that will run untill all compaction test cases are run
-	etcd, err = utils.StartEmbeddedEtcd(testCtx, testEtcdDir, logger)
+	etcd, err = utils.StartEmbeddedEtcd(testCtx, testEtcdDir, logger, embeddedEtcdPortNo)
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/compactor/compactor_suite_test.go:72]: The issue which I faced is that [StartEmbeddedEtcd](https://github.com/gardener/etcd-backup-restore/blob/master/test/utils/utils.go#L46) starting at [random port](https://github.com/gardener/etcd-backup-restore/blob/master/test/utils/utils.go#L54-L57) but AFAIK that port is not exposed to fetch the endpoints correctly. Although `compact pkg` unit test will get passed because they are not throwing an error if defragmentation fails but you look closely you can find the logs.
```
time="2021-11-04T18:03:39+05:30" level=info msg="Starting the defragmentation on etcd leader" suite=compactor
time="2021-11-04T18:03:39+05:30" level=info msg="Defragmenting etcd member[http://localhost:0]" suite=compactor
.
.
.
time="2021-11-04T18:02:24+05:30" level=warning msg="Failed to get status of etcd member[http://localhost:0] with error: context deadline exceeded" suite=compactor
```

And  I got this error while running defragmentation unit tests:
```
{"level":"warn","ts":"2021-11-04T17:22:12.796+0530","caller":"grpclog/grpclog.go:60","msg":"grpc: addrConn.createTransport failed to connect to {http://localhost:0  <nil> 0 <nil>}. Err :connection error: desc = \"transport: Error while dialing dial tcp [::1]:0: connect: can't assign requested address\". Reconnecting..."}
``` 
Amshu suggested me to StartEmbeddedEtcd on particular port, I did that but then `compact pkg` throw an error as it is using to 2 embedded-etcd during testing, then I thought better to make portNo. of StartEmbeddedEtcd configurable.

File: pkg/initializer/validator/datavalidator_test.go [hunk 15]
```diff
@@ -209,7 +212,7 @@ var _ = Describe("Running Datavalidator", func() {
 			}()
 
 			// start etcd
-			etcd, err := utils.StartEmbeddedEtcd(testCtx, restoreDataDir, logger)
+			etcd, err := utils.StartEmbeddedEtcd(testCtx, restoreDataDir, logger, embeddedEtcdPortNo)
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/initializer/validator/datavalidator_test.go:215]: we need  `embeddedEtcdPortNo` specially  for `compact pkg` and `defragmentor pkg` unit tests. Here I mentioned this just for feature completion

File: pkg/leaderelection/leaderelection.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (amshuman-kr) [reply pkg/leaderelection/leaderelection.go:135]: Passing different config to different sidecars will be hard given that we have only one `StatefulSet` which creates all the pods of the ETCD cluster.

The member sidecars need to still distinguish between themselves and this can be achieved by passing the pod name (which would also be the ETCD member name according to the proposal) as an environment variable using `fieldRef` mechanism.

The code here should ideally either iterate through the member names and select the correct endpoint of the member by comparing with the pod/member name passed via env variable or it should rely on some convention in configuration where the endpoint can be computed based on the passed pod/member name.

**Comment by ishan16696** [comment]
Rebased and Resolve the merge conflicts.

File: pkg/server/backuprestoreserver.go [hunk 16]
```diff
@@ -128,67 +138,105 @@ func (b *BackupRestoreServer) startHTTPServer(initializer initializer.Initialize
...
-	go defragmentor.DefragDataPeriodically(ctx, b.config.EtcdConnectionConfig, b.defragmentationSchedule, ssr.TriggerFullSnapshot, b.logger)
+	if runServerWithSnapshotter {
+		go handleAckState(handler, ackCh)
+	}
 
 	//TODO @aaronfern: Add functionality for member garbage collection
 	if b.config.HealthConfig.MemberLeaseRenewalEnabled {
-		go heartbeat.RenewMemberLeasePeriodically(ctx, b.config.HealthConfig, b.logger, b.config.EtcdConnectionConfig)
+		go heartbeat.RenewMemberLeasePeriodically(ctx, mmStopCh, b.config.HealthConfig, b.logger, b.config.EtcdConnectionConfig)
 	}
```
Comment (aaronfern) [inline pkg/server/backuprestoreserver.go:232]: Why is this required if `memberLeaseCallbacks` already does the same task?

**Review by aaronfern — COMMENTED** [review]
Just one query from me @ishan16696 
Thanks

File: pkg/server/backuprestoreserver.go (hunk 16 above)
Comment (ishan16696) [reply pkg/server/backuprestoreserver.go:232]: Actually all backup-restore have default state as `Follower`, hence `RenewMemberLeasePeriodically()` starts for everybody  and `memberLeaseCallbacks` is going to take care when there is any state change(refer [doc](https://github.com/gardener/etcd-backup-restore/blob/eb57e66f058b3176c2720a5504e84e91fb1e2cf5/doc/concepts/leader_election.md#complete-work-flow-leader-election-state-diagram)) in backup-restore.

Comment (aaronfern) [reply pkg/server/backuprestoreserver.go:232]: Got it. Thanks for explaining 👍🏻

File: pkg/leaderelection/types.go [hunk 17]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
```
Comment (abdasgupta) [inline pkg/leaderelection/types.go:1]: Can we move this file to pkg/types directory? That directory contains all the types

**Review by abdasgupta — CHANGES_REQUESTED** [review]
I had an initial look. Overall, great effort. Thanks.

File: pkg/leaderelection/leaderelection.go [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+				// start the snapshotter.
+				le.CurrentState = StateLeader
+				le.logger.Infof("backup-restore became: %v", le.CurrentState)
+
+				if le.Callbacks.OnStartedLeading != nil {
+					leCtx, leCancel = context.WithCancel(ctx)
+					le.logger.Info("backup-restore started leading...")
+					le.Callbacks.OnStartedLeading(leCtx)
+				}
+			} else if isLeader && le.CurrentState == StateLeader {
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:97]: yes here I'm not taking any action just log in debug level `le.logger.Debug("no change in leadershipStatus...")`

File: pkg/leaderelection/leaderelection.go [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+				le.logger.Infof("backup-restore is in: %v", le.CurrentState)
+				if le.Callbacks.OnStoppedLeading != nil && leCtx != nil {
+					leCancel()
+					le.Callbacks.OnStoppedLeading()
+				}
+				le.logger.Info("waiting for Re-election...")
+				continue
+			}
+
+			if isLeader && (le.CurrentState == StateFollower || le.CurrentState == StateUnknown || le.CurrentState == StateCandidate) {
```
[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:80]: Basically I used `le.CurrentState` as previous state to determine when to action and what action. (refer [document](https://github.com/gardener/etcd-backup-restore/blob/eb57e66f058b3176c2720a5504e84e91fb1e2cf5/doc/concepts/leader_election.md#complete-work-flow-leader-election-state-diagram))

[the remark this answers is not in the document]
Comment (ishan16696) [reply pkg/leaderelection/leaderelection.go:80]: If we don’t use the `le.CurrentState` then how to determine whether it is a new Leader or it is still the same leader(no change in leadership status).
If it is the new Leader then we will take certain action like startSnapshotter etc.
If it is same leader then no action needed. Hope that make sense

File: pkg/leaderelection/types.go (hunk 17 above)
Comment (ishan16696) [reply pkg/leaderelection/types.go]: done

File: pkg/leaderelection/leaderelection.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (abdasgupta) [reply pkg/leaderelection/leaderelection.go:107]: Okay

**Comment by abdasgupta** [comment]
\lgtm

**Review by abdasgupta — APPROVED** [review]
